### ISSUE TITLE
Unboxed Tuples

### ISSUE THREAD

**Issue by gridaphobe** [ucsd-progsys/liquidhaskell#718]
Unboxed Tuples crash LH at the moment.

http://goto.ucsd.edu:8090/index.html#?demo=permalink%2F1464133143_12050.hs

```
LiquidHaskell Copyright 2009-15 Regents of the University of California. All Rights Reserved.

[1;94m
**** DONE:  Parsed All Specifications ******************************************
[0m[1;90m 
[0mliquid: Error: bytecode compiler can't handle unboxed tuples.
  Possibly due to foreign import/export decls in source.
  Workaround: use -fobject-code, or compile this module to .o separately.
```

This seems to be a GHC error that arises because we set `HscInterpreted` in

https://github.com/ucsd-progsys/liquidhaskell/blob/master/src/Language/Haskell/Liquid/GHC/Interface.hs#L123-L126

The FIXME says we need this in order to lookup certain unexported Names in modules.

1. Do we still need this in light of @spinda's work on `Bare`?
2. If not, can we use `-fobject-code` instead to support unboxed tuples?

**Comment by gridaphobe** [comment]
@thoughtpolice points out that there's also a parse error when you try to give a spec to an unboxed tuple, eg

```
{-@ newByteArray# :: Int# -> State# s -> (#State# s, MutableByteArray# s#) @-}
```

This ought to be easy to fix though.

**Comment by spinda** [comment]
Is there any downside to setting `-fobject-code` for now, as a temporary solution?

**Comment by gridaphobe** [comment]
Yes, it doesn't work :( 

If I change `HscInterpreted` to `HscAsm` I get the following

```
> stack exec liquid -- tests/pos/Class2.hs -i tests/neg
LiquidHaskell Copyright 2009-15 Regents of the University of California. All Rights Reserved.

liquid: Cannot add module Class2 to context: not interpreted
```

Removing the call to `setContext` (which is causing the above error) results in

```
> stack exec liquid -- tests/pos/Class2.hs -i tests/neg
LiquidHaskell Copyright 2009-15 Regents of the University of California. All Rights Reserved.


**** DONE:  annotate ***********************************************************


**** RESULT: ERROR *************************************************************


 /Users/gridaphobe/Source/liquid/haskell/.stack-work/install/x86_64-osx/lts-5.9/7.10.3/share/x86_64-osx-ghc-7.10.3/liquidhaskell-0.6.0.0/include/Prelude.spec:43:20: Error: GHC Error

 43 | type FF      = {v: Bool          | not (Prop v)}
                         ^

     Not in scope: type constructor or class `Bool'
```

So, as you can see, we really depend on the behavior of `HscInterpreted`. More specifically, I think we depend on the behavior of [`targetRetainsAllBindings`](https://downloads.haskell.org/~ghc/latest/docs/html/libraries/ghc-8.0.1/DynFlags.html#v:targetRetainsAllBindings) which determines whether GHC keeps the `GlobalRdrEnv` around so we can lookup arbitrary `Name`s.

In theory we should be able to use `-fno-code` (`HscNothing`) since we don't actually care about getting anything beyond the Core, but in the past this has always caused problems for resolving specs for **imported** modules. I've thrown another build on circleci just to see what happens.. https://circleci.com/gh/ucsd-progsys/liquidhaskell/578

**Comment by gridaphobe** [comment]
@spinda Interestingly, setting `HscNothing` seems to work now for all of the tests except when TH is used (which is to be expected). If you can sort out how to selectively enable `HscInterpreted` for modules that use TH (I think there's an API function for this), we should be able to make the switch.

**Comment by spinda** [comment]
There's `mi_used_th` and `mg_used_th`, but both of those only work after a module is compiled. We could just scan the `ParsedModule` for the presence of any splices, and then switch on `HscInterpreted` if TH is detected. Does that sound okay @gridaphobe?

**Comment by gridaphobe** [comment]
Here's the function I was thinking of.

https://downloads.haskell.org/~ghc/latest/docs/html/libraries/ghc-8.0.1/GHC.html#v:needsTemplateHaskell

**Comment by spinda** [comment]
@gridaphobe Implemented in #722. What about the case a module uses both TH and unboxed tuples?

**Comment by ranjitjhala** [comment]
Fixed in develop by PR #722
